src/dfm/common/data_cleaning/dolma_taggers/perplexity.py

```diff
@@ -0,0 +1,180 @@
+            with Path.open(file_path, "wb") as file:
+                file.write(response.content)
+            logging.info(f"{lang} model downloaded and saved at {file_path}")
+        else:
+            raise RuntimeError(
+                f"Failed to download {lang} model. Status code: {response.status_code}",
+            )
+    else:
+        logging.info(f"{lang} model already exists at {file_path}")
+
+    return file_path
+
+
+def pp(log_score: float, length: float) -> float:
+    """Convert total log-probability to perplexity"""
+    return 10.0 ** (-log_score / length)
+
+
+class PerplexityBaseTagger(BaseTagger):
+    @property
+    def model(self: Self) -> kenlm.Model:
+        return self._model
+
+    @model.setter
+    def model(self: Self, model: kenlm.Model):
+        self._model = model
+
+
+def create_ccnet_perplexity_tagger(lang: str) -> type[PerplexityBaseTagger]:
+    """Dynamically create tagger class for a given language"""
+
+    def __init__(self: Any) -> None:
+        model_bin_path = _get_ccnet_pretrained_lm(lang)
+        self.model = kenlm.Model(str(model_bin_path))
+
+    def predict(self: PerplexityBaseTagger, doc: Document) -> DocResult:
+        paragraphs = split_paragraphs(doc.text)
+        spans: list[Span] = []
+        doc_log_prob: float = 0.0
+        doc_length: float = 0.0
+        for paragraph in paragraphs:
+            # To get proper scores from the language model we need to normalize the text
+            # Do not remove accents as it removes æøå and others.
+            normalized_text = blingfire.normalize_spaces(
+                normalize(paragraph.text, accent=False),
+            )
+            # The kenlm model expects end of sentence punctuation to be separated from words with spaces
+            # so we separate the words using blingfire.
+            normalized_words = blingfire.text_to_words(normalized_text)
+            log_prob = self.model.score(normalized_words)
+            length = len(normalized_words.split()) + 1
+            doc_log_prob += log_prob
+            doc_length += length
+            paragraph_span = Span(
+                start=paragraph.start,
+                end=paragraph.end,
+                type="perplexity",
+                score=pp(log_prob, length),
+            )
+            spans.append(paragraph_span)
+
+        paragraph_span = Span(
+            start=0,
+            end=len(doc.text),
+            type="doc_perplexity",
+            score=pp(doc_log_prob, doc_length),
+        )
+        return DocResult(doc=doc, spans=spans)
+
+    # Build the class dynamiccaly from base class
+    # and methods.
+    cls = type(
+        f"CCNetPerplexity{lang}",
+        (PerplexityBaseTagger,),
+        {
+            "__init__": __init__,
+            "predict": predict,
+        },
+    )
+    # Add the class decorator explicitly to add the tagger to the registry
+    cls = TaggerRegistry.add(f"ccnet_perplexity_paragraph_w_doc_{lang}")(cls)
+    return cls
```

Comment:
I'm insufficiently proficient in how the registry works, but I assume you've tested this 👍

Reply:
Yes, but it's a good point that I should add some tests that at least sanity checks the taggers.